### ISSUE TITLE
I tryed to train conformer-rnn-transducer on the CSJ dataset.

### ISSUE THREAD

**Issue by YutaTakase** [espnet/espnet#4813]
I created a confomer-rnn-trasducer recipe using CSJ and run the training process.
When I evaluated the created model, the CERs for eval1, eval2, and eval3 were higher than the [original paper](https://arxiv.org/pdf/2106.05111.pdf).
My recipe was created with [aishell2's conformer-rnn-transducer](https://github.com/espnet/espnet/tree/2b009859b5aa728c8682f372926865726fca48cd/egs2/aishell2/asr1/conf/tuning/transducer) as a reference.

I am wondering if there is something wrong with the conformer-rnn-trasducer recipe I created for the CSJ dataset.
Could you please give me some hints?
Best regards.

||Original CER| My model CER|
|--------| :----: | :---: |
| eval1  | 4.1  | 6.5 |
| eval2  | 3.2  | 4.9 |
| eval3  | 3.5  | 5.5 |

confomer-rnn-trasducer.yaml

```yaml
batch_type: numel
batch_bins: 500000
accum_grad: 4
max_epoch: 50
patience: none
init: none
num_att_plot: 0

val_scheduler_criterion:
    - valid
    - loss
best_model_criterion:
-   - valid
    - cer_transducer
    - min
keep_nbest_models: 10

model_conf:
    ctc_weight: 0.3
    report_cer: True
    report_wer: True

encoder: conformer
encoder_conf:
    output_size: 512
    attention_heads: 8
    linear_units: 2048
    num_blocks: 12
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    attention_dropout_rate: 0.1
    input_layer: conv2d
    normalize_before: true
    macaron_style: true
    rel_pos_type: latest
    pos_enc_layer_type: "rel_pos"
    selfattention_layer_type: "rel_selfattn"
    activation_type: "swish"
    use_cnn_module:  true
    cnn_module_kernel: 15

decoder: transducer
decoder_conf:
    rnn_type: lstm
    num_layers: 1
    hidden_size: 512
    dropout: 0.1
    dropout_embed: 0.2

joint_net_conf:
    joint_space_size: 512

optim: adam
optim_conf:
    lr: 0.001
    weight_decay: 0.000001
scheduler: warmuplr
scheduler_conf:
    warmup_steps: 30000

frontend_conf:
  n_fft: 512
  hop_length: 160

specaug: specaug
specaug_conf:
    apply_time_warp: true
    time_warp_window: 5
    time_warp_mode: bicubic
    apply_freq_mask: true
    freq_mask_width_range:
    - 0
    - 30
    num_freq_mask: 2
    apply_time_mask: true
    time_mask_width_range:
    - 0
    - 40
    num_time_mask: 2
```

decode.yaml

```yaml
beam_size: 10
lm_weight: 0.2
transducer_conf:
    search_type: default
    score_norm: False
```

Train and Validation loss.
After 20 epochs, Training process was stopped because the output was "No improvement in this epoch".

```txt
1epoch results: [train] iter_time=0.026, forward_time=0.032, loss_ctc=67.830, loss_transducer=56.493, loss=76.842, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.143e-04, train_time=0.817, time=4 hours, 16 minutes and 46.45 seconds, total_count=75423, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=26.945, cer_ctc=nan, loss_transducer=16.946, cer_transducer=0.170, wer_transducer=0.805, loss=25.030, time=3 minutes and 50.5 seconds, total_count=811, gpu_max_cached_mem_GB=13.770
2epoch results: [train] iter_time=0.031, forward_time=0.032, loss_ctc=34.722, loss_transducer=21.739, loss=32.155, backward_time=0.042, optim_step_time=0.037, optim0_lr0=8.669e-04, train_time=0.877, time=4 hours, 35 minutes and 38.77 seconds, total_count=150846, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=19.159, cer_ctc=nan, loss_transducer=12.291, cer_transducer=0.133, wer_transducer=0.757, loss=18.039, time=3 minutes and 56.33 seconds, total_count=1622, gpu_max_cached_mem_GB=13.770
3epoch results: [train] iter_time=0.020, forward_time=0.032, loss_ctc=26.163, loss_transducer=16.195, loss=24.044, backward_time=0.042, optim_step_time=0.038, optim0_lr0=8.018e-04, train_time=0.784, time=4 hours, 6 minutes and 28.69 seconds, total_count=226269, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=14.210, cer_ctc=nan, loss_transducer=9.121, cer_transducer=0.109, wer_transducer=0.710, loss=13.384, time=3 minutes and 41.14 seconds, total_count=2433, gpu_max_cached_mem_GB=13.770
4epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=22.152, loss_transducer=13.517, loss=20.163, backward_time=0.042, optim_step_time=0.037, optim0_lr0=6.760e-04, train_time=0.610, time=3 hours, 11 minutes and 59.5 seconds, total_count=301692, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=12.289, cer_ctc=nan, loss_transducer=8.141, cer_transducer=0.095, wer_transducer=0.671, loss=11.828, time=3 minutes and 24.72 seconds, total_count=3244, gpu_max_cached_mem_GB=13.770
5epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=19.838, loss_transducer=11.949, loss=17.901, backward_time=0.042, optim_step_time=0.036, optim0_lr0=5.955e-04, train_time=0.604, time=3 hours, 10 minutes and 4.6 seconds, total_count=377115, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=10.921, cer_ctc=nan, loss_transducer=7.352, cer_transducer=0.088, wer_transducer=0.654, loss=10.628, time=3 minutes and 28.21 seconds, total_count=4055, gpu_max_cached_mem_GB=13.770
6epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=18.316, loss_transducer=10.912, loss=16.407, backward_time=0.041, optim_step_time=0.037, optim0_lr0=5.384e-04, train_time=0.606, time=3 hours, 10 minutes and 33.14 seconds, total_count=452538, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=9.482, cer_ctc=nan, loss_transducer=6.395, cer_transducer=0.078, wer_transducer=0.617, loss=9.240, time=3 minutes and 39.17 seconds, total_count=4866, gpu_max_cached_mem_GB=13.770
7epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=17.181, loss_transducer=10.153, loss=15.307, backward_time=0.042, optim_step_time=0.037, optim0_lr0=4.951e-04, train_time=0.608, time=3 hours, 11 minutes and 16.97 seconds, total_count=527961, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=9.870, cer_ctc=nan, loss_transducer=6.520, cer_transducer=0.083, wer_transducer=0.629, loss=9.481, time=3 minutes and 24.54 seconds, total_count=5677, gpu_max_cached_mem_GB=13.770
8epoch results: [train] iter_time=0.003, forward_time=0.032, loss_ctc=16.270, loss_transducer=9.544, loss=14.425, backward_time=0.042, optim_step_time=0.037, optim0_lr0=4.608e-04, train_time=0.618, time=3 hours, 14 minutes and 26.73 seconds, total_count=603384, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=9.242, cer_ctc=nan, loss_transducer=6.179, cer_transducer=0.074, wer_transducer=0.611, loss=8.952, time=3 minutes and 30.88 seconds, total_count=6488, gpu_max_cached_mem_GB=13.770
9epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=15.535, loss_transducer=9.052, loss=13.712, backward_time=0.042, optim_step_time=0.037, optim0_lr0=4.328e-04, train_time=0.609, time=3 hours, 11 minutes and 40.7 seconds, total_count=678807, gpu_max_cached_mem_GB=13.770, [valid] loss_ctc=8.754, cer_ctc=nan, loss_transducer=6.012, cer_transducer=0.075, wer_transducer=0.615, loss=8.638, time=3 minutes and 33.44 seconds, total_count=7299, gpu_max_cached_mem_GB=13.770
10epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=14.934, loss_transducer=8.649, loss=13.129, backward_time=0.042, optim_step_time=0.037, optim0_lr0=4.094e-04, train_time=0.601, time=3 hours, 9 minutes and 6.36 seconds, total_count=754230, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=8.668, cer_ctc=nan, loss_transducer=5.861, cer_transducer=0.072, wer_transducer=0.602, loss=8.461, time=3 minutes and 25.96 seconds, total_count=8110, gpu_max_cached_mem_GB=13.771
11epoch results: [train] iter_time=0.002, forward_time=0.033, loss_ctc=14.409, loss_transducer=8.302, loss=12.624, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.894e-04, train_time=0.614, time=3 hours, 13 minutes and 14.36 seconds, total_count=829653, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=8.265, cer_ctc=nan, loss_transducer=5.611, cer_transducer=0.067, wer_transducer=0.582, loss=8.090, time=3 minutes and 23.16 seconds, total_count=8921, gpu_max_cached_mem_GB=13.771
12epoch results: [train] iter_time=0.006, forward_time=0.033, loss_ctc=13.966, loss_transducer=8.004, loss=12.193, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.720e-04, train_time=0.637, time=3 hours, 20 minutes and 26.1 seconds, total_count=905076, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=8.354, cer_ctc=nan, loss_transducer=5.630, cer_transducer=0.069, wer_transducer=0.588, loss=8.136, time=3 minutes and 39.52 seconds, total_count=9732, gpu_max_cached_mem_GB=13.771
13epoch results: [train] iter_time=0.005, forward_time=0.032, loss_ctc=13.565, loss_transducer=7.745, loss=11.814, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.568e-04, train_time=0.625, time=3 hours, 16 minutes and 26.85 seconds, total_count=980499, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.946, cer_ctc=nan, loss_transducer=5.462, cer_transducer=0.068, wer_transducer=0.579, loss=7.845, time=3 minutes and 28.39 seconds, total_count=10543, gpu_max_cached_mem_GB=13.771
14epoch results: [train] iter_time=0.002, forward_time=0.032, loss_ctc=13.218, loss_transducer=7.519, loss=11.485, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.434e-04, train_time=0.613, time=3 hours, 12 minutes and 41.47 seconds, total_count=1055922, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=8.238, cer_ctc=nan, loss_transducer=5.672, cer_transducer=0.067, wer_transducer=0.582, loss=8.144, time=3 minutes and 46.53 seconds, total_count=11354, gpu_max_cached_mem_GB=13.771
15epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=12.846, loss_transducer=7.277, loss=11.131, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.313e-04, train_time=0.600, time=3 hours, 8 minutes and 52.21 seconds, total_count=1131345, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.613, cer_ctc=nan, loss_transducer=5.210, cer_transducer=0.061, wer_transducer=0.551, loss=7.494, time=3 minutes and 29.86 seconds, total_count=12165, gpu_max_cached_mem_GB=13.771
16epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=12.560, loss_transducer=7.093, loss=10.861, backward_time=0.041, optim_step_time=0.037, optim0_lr0=3.204e-04, train_time=0.603, time=3 hours, 9 minutes and 30.31 seconds, total_count=1206768, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.292, cer_ctc=nan, loss_transducer=5.073, cer_transducer=0.063, wer_transducer=0.564, loss=7.260, time=3 minutes and 24.42 seconds, total_count=12976, gpu_max_cached_mem_GB=13.771
17epoch results: [train] iter_time=0.002, forward_time=0.033, loss_ctc=12.327, loss_transducer=6.942, loss=10.640, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.106e-04, train_time=0.605, time=3 hours, 10 minutes and 17.22 seconds, total_count=1282191, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.748, cer_ctc=nan, loss_transducer=5.320, cer_transducer=0.066, wer_transducer=0.575, loss=7.644, time=3 minutes and 24.02 seconds, total_count=13787, gpu_max_cached_mem_GB=13.771
18epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=12.099, loss_transducer=6.796, loss=10.426, backward_time=0.042, optim_step_time=0.037, optim0_lr0=3.016e-04, train_time=0.602, time=3 hours, 9 minutes and 28.86 seconds, total_count=1357614, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.794, cer_ctc=nan, loss_transducer=5.369, cer_transducer=0.067, wer_transducer=0.576, loss=7.707, time=3 minutes and 25.91 seconds, total_count=14598, gpu_max_cached_mem_GB=13.771
19epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=11.875, loss_transducer=6.650, loss=10.212, backward_time=0.041, optim_step_time=0.037, optim0_lr0=2.933e-04, train_time=0.600, time=3 hours, 8 minutes and 41.07 seconds, total_count=1433037, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.173, cer_ctc=nan, loss_transducer=5.031, cer_transducer=0.062, wer_transducer=0.553, loss=7.183, time=3 minutes and 36.95 seconds, total_count=15409, gpu_max_cached_mem_GB=13.771
20epoch results: [train] iter_time=0.001, forward_time=0.032, loss_ctc=11.619, loss_transducer=6.500, loss=9.986, backward_time=0.042, optim_step_time=0.037, optim0_lr0=2.857e-04, train_time=0.606, time=3 hours, 10 minutes and 38.84 seconds, total_count=1508460, gpu_max_cached_mem_GB=13.771, [valid] loss_ctc=7.212, cer_ctc=nan, loss_transducer=5.035, cer_transducer=0.062, wer_transducer=0.555, loss=7.199, time=3 minutes and 42.97 seconds, total_count=16220, gpu_max_cached_mem_GB=13.771
```

My environment is as follows.

```txt
PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.26

Python version: 3.7.13 (default, Nov 14 2022, 01:47:36)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-1022-aws-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 515.65.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] pytorch-ranger==0.1.1
[pip3] pytorch-wpe==0.0.1
[pip3] torch==1.12.1
[pip3] torch-complex==0.4.3
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==0.12.1
[pip3] warprnnt-pytorch==0.1
[pip3] espnet==202209
```

**Comment by sw005320** [comment]
> After 20 epochs, Training process was stopped because the output was "No improvement in this epoch".

It is most likely because it stopped at 20 epochs.
Note that "No improvement in this epoch" is just a message. Not an error.
https://github.com/espnet/espnet/blob/master/espnet2/train/trainer.py#L376

I'm expecting the other error happens during your training.
Please check the various logs carefully.

**Comment by YutaTakase** [comment]
Thank you for your response.
Sorry, I misunderstood "No improvement in this epoch" as an error or warning.
I will check the various logs and try training process over 20epochs.

**Comment by YutaTakase** [comment]
Thanks for the advice.
As advised, accuracy improved after continued training.
I have training and evaluated up to 70 epochs and the CER is as follows.
|| Original CER| My model CER|
| ----- | --- | --- |
| eval1 | 4.1 | 4.8 |
| eval2 | 3.2 | 3.6 |
| eval3 | 3.5 | 4.0 |

Reference
|cer_transducer vs epoch|loss_transducer vs epoch|
|-----------------------|------------------------|
|![](https://user-images.githubusercontent.com/103028291/212799367-041b76b7-aa4a-4f81-b240-215ee72d1d17.png)|![](https://user-images.githubusercontent.com/103028291/212799416-5994ec63-229b-4b12-b91b-a0c766a6ac6d.png)|


I checked the number of parameters in the transducer model and it was 90.8M.

On the other hand, the number of parameters listed in the paper was 120M, so we believe that the slightly different configuration of the model may have affected the accuracy.

I would like to revise the configuration of the model in future additional experiments.

**Comment by sw005320** [comment]
Very cool!
Please report the progress.
This is very informative.
Thanks for sharing your setup.
